### ISSUE TITLE
Font-Size different (LARGE) in Azure Web App

### ISSUE THREAD

**Issue by jsheely** [Automattic/node-canvas#857]
When I deploy a working example to my Azure Web App instance the font-size at 14px is GIGANTIC (like 100x). I have to set the font size to 1px to be able to see more than part of one letter. 

Any idea what causes this inconsistency between environments? They are both: 

- Windows 64-bit
- Node 6.9.1 (64-bit)
- Canvas 1.6.2
- Prebuilt canvas.node (in git repo)

### Examples

**Working Localhost (Size 20px Font)**
![image](https://cloud.githubusercontent.com/assets/1424946/21146448/af03de8c-c11f-11e6-893c-3a4461d37997.png)

**Busted Azure Web App (Size 1px Font)**
![image](https://cloud.githubusercontent.com/assets/1424946/21146455/b3f98e78-c11f-11e6-8c42-81bc0c4d3c94.png)

## Steps to Reproduce

```js
var Canvas = require('canvas');
var canvas = new Canvas(400,100);
var ctx = canvas.getContext('2d');

const fontSize = process.env.FONT_SIZE || '20';

ctx.fillStyle = 'black';
ctx.font = `${fontSize}px Calibri`;
ctx.fillText("Hello World", 105, 30);
```

**Comment by chearon** [comment]
Do you mean you're using my canvas-prebuilt from [here](https://github.com/chearon/node-canvas-prebuilt)? And I assume you're using the exact same prebuilt locally right? I can't think of anything that would cause something like that other than maybe a wrong PANGO_SCALE

**Comment by jsheely** [comment]
@chearon Oh, I actually meant that I deployed the build artifacts from localhost. Just to be clear that the server wasn't building anything itself (because it can't). I did also try the canvas-prebuilt just in case but same result.

It's my understanding that PANGO isn't even enabled in windows. *bindings.gyp* seems to suggest just that with the *with_pango: false*.

```js
['OS=="win"', {
      'variables': {
        'GTK_Root%': 'C:/GTK', # Set the location of GTK all-in-one bundle
        'with_jpeg%': 'false',
        'with_gif%': 'false',
        'with_pango%': 'false',
        'with_freetype%': 'false'
      }
...
```

**Comment by chearon** [comment]
Yeah, you're right, for 1.6 Pango is disabled on Windows (is required for the unreleased 2.0 though). The font system used by 1.6 is really limited... Is using canvas@master a possibility? If you're not doing custom fonts it should work just the same but it will use Pango to render fonts instead of Cairo

**Comment by jsheely** [comment]
@chearon Unfortunately the same result using canvas@master

If master is 2.0 then why is package.json still set to version 1.6.0 (which is even older than the npm install canvas of 1.6.2)

Am I missing something?

**Comment by chearon** [comment]
Nah, that's just because the version gets bumped at the point of the release which hasn't happened yet. 1.6.2 is on a different branch which was created just before the stuff for 2.0 started being merged into master.

I'm really not sure about the font size thing though, can you try a different font and see if you get the same results?

**Comment by jsheely** [comment]
@chearon Okay great. Just making sure

I did try:
```js
ctx.font = '14px Arial';
```

and got the same gigantic result. Should I be registering a custom font? I did see a note at the point in the code where font size is determined that there is a TODO for DPI. I don't know a lot about the internals here but could the server have some DPI setting that is different?

**Comment by chearon** [comment]
Yeah, it must be something like that. Except that if both rendering through Cairo directly (1.6) and rendering through Pango->Cairo (master) had the same issue it makes me think that the issue might be somewhere in the Windows font rendering calls...

**Comment by jsheely** [comment]
You're probably right about the Windows font rendering call. Just not sure what would be different on the system to effect it. 

This would be a Cairo issue?

**Comment by chearon** [comment]
If it's an issue with Cairo units then drawing a square or a circle on the canvas should be huge on Azure, too, so maybe try that. That'd make sense but you probably would have noticed that already. 

Only other thing I can think of is that it's somewhere in the Windows font rendering calls, which would take some isolated test programs to narrow it down (is Azure free? If so maybe I can take a stab at debugging it)

**Comment by jsheely** [comment]
@chearon Yes Azure is free for a specific pricing tier. However we need to be under the basic tier to enable 64-bit.

I created a repo https://github.com/jsheely/canvas-azure-fonts

I enabled it to to auto-deploy to an azure instance http://canvas-azure-fonts.azurewebsites.net/Hello%20World

I gave you the ability to modify the repo to trigger the deployment. Feel free to test it out and I'd love to help out in any way.

**Comment by chearon** [comment]
Great! I'll start poking around and let you know if I find anything

**Comment by chearon** [comment]
I seem to have broken your test web app 😳

I saw some hints in the Cairo source (how node-canvas@1.6/Windows renders fonts) that fonts go through WinGDI, and I already know that Pango does that too, so I think that explains why you saw the issue with both versions.

I also tried changing DPI settings on my copy of Windows to see if that caused fonts to scale up, but they didn't. It's really hard to know without having some access or details on how the OS is customized.

**Comment by jsheely** [comment]
Hmm. I just added a line break and redeployed and it seems like it works. Not sure what that was about.

It's worth noting that using asp.net System.Drawing and picking a font-size does not have this issue. But I'm sure they probably take different approaches to actually rendering the font.

I can probably get some settings if I knew what I was looking for.

**Comment by chearon** [comment]
Really?! They should probably end up using the same drawing calls... I guess the second thing I would do is get the example [here](https://developer.gnome.org/pango/stable/pango-Cairo-Rendering.html) to compile + render, and see if it has the same font size problem. And then if it does, I'd check `pango_cairo_font_map_get_resolution(pango_cairo_font_map_get_default())` to see if that number is wrong

**Comment by jsheely** [comment]
C is a little out of my comfort zone but I'll give it a whirl. Would be interesting to know what system setting that ```pango_cairo_font_map_get_default()``` requests. Then I could just look it up on the system

**Comment by jsheely** [comment]
I wonder if because there is no monitor hooked up to this machine and there is no X,Y resolution set for the screen. It's like pulling a MAX INT value or something as the default. Just a random thought

**Comment by chearon** [comment]
I was thinking it has to do with being a headless installation, too, it wouldn't be the first time I found a bug in non-Linux Pango

> Looking at the Pango source I see a switch for Win32 vs FreeType

Yeah that's where it picks the backend, on Windows it would pick the win32 (`pangowin32-fontmap.c`) not the FT one, which is for Linux. I'd also look at [this repo](https://github.com/GNOME/pango/blob/master/pango/pangocairo-fontmap.c) since it's changed a lot since whenever 

I think [here](https://github.com/GNOME/pango/blob/master/pango/pangocairo-win32fontmap.c#L121) is where it sets the DPI

**Comment by jsheely** [comment]
So looking at the resolution method here. I guess I'd have to figure out what what's going on with [this value](https://github.com/GNOME/pango/blob/e9471764d19f740f756cf7e6d11aba0cbba3d2ae/pango/pangowin32-fontmap.c#L629)

Then my next question is knowing that this value is going to be wrong. Is there a way within node-canvas I can force the compensation?

**Comment by jsheely** [comment]
Looking at the MSDN article for [GetDeviceCaps](https://msdn.microsoft.com/en-us/library/windows/desktop/dd144877(v=vs.85).aspx) the ```LOGPIXELSY``` is most likely the strange value on this headless server.

**Comment by chearon** [comment]
Yep, but it should be returning the DPI setting... so normally it should be a value of 96. And that `resolution` should end up being "font units per pixel".

I'd really like to test what that resolution is so that a patch could be submitted to Pango. I got one accepted before... always best to solve the problem at the root! If not, your work-around would be figuring out that value in JS. Hopefully there would be some system command to get the value. Then you'd just, whenever printing fonts, scale by (96 / the value) I think

**Comment by chearon** [comment]
If you can figure out any way to see what that value is, though, that'd be really useful to know. I'm struggling a bit with Azure right now, I can create a program that will print it but then how do I upload it/get it to execute/print the value out in the browser?

**Comment by jsheely** [comment]
I ran run the app if you can create an .exe that will just print the value out

**Comment by chearon** [comment]
Here you go: [res.zip](https://github.com/Automattic/node-canvas/files/655573/res.zip). 64 bit. It has a ton of bundled DLLs cause it has to bring in the entire Pango/Cairo/GLib stack. Mine prints out the expected `96`. I'll wait in anticipation

**Comment by jsheely** [comment]
It returns a blank
![image](https://cloud.githubusercontent.com/assets/1424946/21238793/5d0b360c-c2d2-11e6-948c-b5b2362bda97.png)

**Comment by chearon** [comment]
Is that on Azure or on another windows computer? mine:

<img width="381" alt="screen shot 2016-12-15 at 2 28 06 pm" src="https://cloud.githubusercontent.com/assets/78187/21238919/d5759ace-c2d2-11e6-8690-b929f5ac44fa.png">

**Comment by jsheely** [comment]
That was on Azure

**Comment by chearon** [comment]
I made a more simple program that only calls `GetDeviceCaps` and it too didn't output anything on Azure. I'm still kind of interested in wtf Windows is doing here, if you want to keep debugging contact me at caleb at my username dot net, either way if I find anything out I'll post it to this issue

**Comment by chearon** [comment]
@zbjornson I know you have some Windows experience, have you ever seen anything like this before?

**Comment by jsheely** [comment]
I was looking at some other repositories and it seems like if they don't get a value back from GetDeviceCaps they just return 96. Seems like that would resolve the issue in pango. 

I don't know how C deals with getting back what would ordinarily be a null when it does the math. I would have assumed an error but evidently it just returns a giant number

**Comment by zbjornson** [comment]
I think that's a limitation of the Hyper-V display driver. The newer `GetDpiForSystem` call falls back to 96 DPI (same as what @jsheely found in other libs using `GetDeviceCaps`); I would take that approach if `GetDeviceCaps` returns a bad value.

**Comment by chearon** [comment]
Well I just got a test program to output the `GetDeviceCaps(LOGPIXELSY)` call on Azure (I think the exe from before was the wrong architecture) and it returns 96 correctly. The Pango resolution also returns 96. Feeling kind of helpless on this now

**Comment by Killian22** [comment]
We found this issue as well and are also feeling helpless.  
We recently are trying to find a magic number to set our font to as a work around but we can't even get that right!

Notice the huge difference between 1.49 and 1.50 ! 

Does anyone know of a workaround ? 

![image](https://user-images.githubusercontent.com/332156/51045431-8c4fec00-1591-11e9-87a0-8a4afd9c9610.png)

**Comment by LinusU** [comment]
possibly related: #99

**Comment by Killian22** [comment]
It seems that it was doing that.  This still leaves us without the ability to make the text the correct size :/ .

**Comment by Killian22** [comment]
We're working on a workaround now where we use a second canvas to draw the text on and then drawImage to scale the image to an exact size. Inital test worked, figuring out the numbers needed to scale correctly now.

**Comment by jorgecf** [comment]
Anyone has a clue? For now, we're deploying our node server on linux where fonts work perfectly, but it'd be nice to have a Windows solution
